src/combinatorics/ramsey/infinite.lean

```diff
@@ -0,0 +1,483 @@
+* Sᵢ₊₁ ⊆ Sᵢ
+* xᵢ₊₁ ∈ Sᵢ
+* xᵢ < xᵢ₊₁
+* f {xᵢ₊₁, y} = f {xᵢ₊₁, z} for all xᵢ₊₁ < y < z in Sₓ, x = xᵢ
+
+Then the set {xᵢ | i ∈ ℕ} has the property that
+f {xᵢ, xⱼ} = f {xᵢ, xₖ} for all i < j < k.
+If we color singletons in this set by f' xᵢ = f {xᵢ, xᵢ₊₁},
+then by the pigeon hole principle there is an infinite subset
+of the xᵢ's whose unordered pairs are all colored the same.
+
+## Notation
+
+This file uses `[S]²` to denote the set of unordered pairs of
+the given set `S`.
+
+## Implementation notes
+
+The set of two colors is defined as an emumeration.
+This definition is easy to work with, and easy to read,
+but comes at the cost of not generalizing.
+
+Unordered pairs are implemented as ordered pairs with the condition
+that the second element is strictly greater than the first.
+This definition is also easy to work with but does not generalize.
+
+## References
+
+*  [Graham, R.L. and Rothschild, B.L. and Spencer, J.H., *Ramsey Theory*][graham1990]
+-/
+
+open set
+open classical
+open nat
+
+section infinite_ramsey_pairs
+
+/--
+A set of natural numbers H is infinite if for any natural
+number there is a larger number in H.
+-/
+def infinite (H : set ℕ) := ∀ x : ℕ, ∃ y : ℕ, x < y ∧ y ∈ H
+
+/--
+There are two colors, red and blue.
+-/
+inductive color
+| red : color
+| blue : color
+
+/--
+An infinite set of natural numbers.
+-/
+structure Inf :=
+(s : set ℕ)
+(pf : infinite s)
+
+instance : has_coe Inf (set ℕ) := ⟨Inf.s⟩
+instance : has_mem ℕ Inf := ⟨λ n H, n ∈ H.s⟩
+instance : has_subset Inf := ⟨λ H₁ H₂, H₁.s ⊆ H₂.s⟩
+
+open color
+
+/--
+Given a function from natural numbers to colors,
+this property describes the numbers which are mapped to red.
+-/
+def reds (f : ℕ → color) (H : set ℕ) := (λ n : ℕ, f n = red ∧ n ∈ H )
+
+/--
+Given a function from natural numbers to colors,
+this property describes the numbers which are mapped to blues.
+-/
+def blues (f : ℕ → color) (H : set ℕ) := (λ n : ℕ, f n = blue ∧ H n)
+
+lemma lt_succ_sum : forall x y : ℕ, x < x + y + 1 :=
+λ x y : ℕ, lt_add_of_pos_right x (succ_pos y)
+
+/--
+Unordered pairs are defined as ordered pairs with the condition
+that the second element is strictly greater than the first.
+-/
+def unordered_pairs (S : set ℕ) :=
+  {p : ℕ × ℕ // p.fst < p.snd ∧ S p.fst ∧ S p.snd}
+
+notation `[ℕ]²` := unordered_pairs univ
+notation `[` S `]²` := unordered_pairs S
+
+/--
+Given a number m and a coloring f, the projection of a coloring
+of unordered_pairs to a coloring of numbers is given by fixing
+the first element as m. Note that red is arbitrarily chosen
+as the color assigned to any number below m.
+-/
+def project (f : [ℕ]² → color) (m : ℕ) : ℕ → color :=
+  λ n : ℕ,
+    dite (m < n)
+      (λ h : m < n, f (subtype.mk (m, n) ⟨h, trivial, trivial⟩))
+      (λ _, red)
+
+/--
+The coloring of an unordered pair {a, b}, where a < b,
+agrees with the projection of a applied to b.
+-/
+lemma project_eq (f : [ℕ]² → color) (p : [ℕ]²):
+  f p = project f p.val.fst p.val.snd :=
```

Comment:
I would imagine that you swap the sides of the equality. This would then make a nice simp-lemma. (I.e., tag it with `@[simp]` before the `lemma` keyword.)

Reply:
ah, great idea! I make that change, and it worked, but when I used it I did have to use
`simp [project_eq f p]` and not just `simp [project_eq]`. but maybe that's expected? then again, in the spot where I used it, I had to construct the `p` in place.